Introduce arbitrary orientation for lumped ports

hughcars opened the pull request:
*Description of changes:*
* Introduces the ability to specify the orientation of a lumped port using a 3 vector `[x,y,z]`, in addition to the existing key word `"X"`, `"-Y"`, `"+R"` style interface.
* Introduces algorithms for computing `BoundingBox` and `BoundingBall`. These work by reducing onto the dominant rank all data associated with a marked portion of the mesh, then applying sorting/separation approaches based on a a highly restricted subset of the QuickHull algorithm. 
* The rings example has been changed to apply a π/6 around (0,0,1). This requires regeneration of the baseline data, and changing of the probe example location and data.

*Nuances:*
* The choice to avoid a full QuickHull implementation or Chan's algorithm (generalizes across processors) was based on pragmatism. The upside is relative simplicity (each algorithm requires at most two projections and distance measurements, and some lexicographic sorting) but at the cost of applying a serial bottle neck. This was viewed as not a practical issue however, as generally a) the number of vertices in a port is small and b) most of these vertices are on one rank. If however there was a port that had many vertices, and was uniformly split amongst processors, this would be problematic for these algorithms.
* Though these algorithms will work for 3D objects, *they should not in general be used for large (made up of many vertices) objects*, due to the reduction potentially placing a large quantity of data on the dominant rank (the rank with the largest number of initial vertices).
* Usage of the bounding box algorithms in `lumpedelement.hpp` is to decide upon a length/width of a rectangular port, and outer radius of a coaxial port. 
  * The appropriate specification of the `"+R"` direction in a coaxial port is non ambiguous, however the specification of circumferential or azimuthal forcing would be ambiguous. This ambiguity was not resolved, as at present neither of these forcing types have been needed.

*Issue #, if available:*
Closes https://github.com/awslabs/palace/issues/71

hughcars commented:
Reverting to draft after discussion about bug in the bounding box length calculation.

hughcars commented on test/ref/rings/probe-B.csv:
Regenerating this data requires rotating the coordinate system around (1,1,1) by pi/3. The computed values are documented in the commit message, but repeat here:
```
julia> AngleAxis(pi/3, 1,1,1) * [-6.975740032e-01,        -3.331547360e+00,        +1.022439079e+02]
3-element StaticArraysCore.SVector{3, Float64} with indices SOneTo(3):
  68.80807171786667
 -36.76738354213334
  66.17409836106667

julia> AngleAxis(pi/3, 1,1,1) * [-2.108257694e-03,        -1.094675073e-02,        +1.007018922e+01]
3-element StaticArraysCore.SVector{3, Float64} with indices SOneTo(3):
  6.715702891780666
 -3.3654330789493336
  6.706864398744667
  ```

sebastiangrimberg replied:
From offline discussion: Reverted the rotation to being about (0,0,1) so that the data makes a bit more sense on visual inspection.

hughcars commented on scripts/schema/config/boundaries.json:
```diff
@@ -383,5 +383,25 @@
+      [
+        {
+          "type": "string",
+          "enum": ["R", "X", "Y", "Z", "+R", "+X", "+Y", "+Z", "-R", "-X", "-Y", "-Z",
+                   "r", "x", "y", "z", "+r", "+x", "+y", "+z", "-r", "-x", "-y", "-z"]
```
After discovering this trick, it's probably worth applying to all the other enum inputs for the config files.

sebastiangrimberg commented on palace/utils/geodata.cpp:
```diff
@@ -61,7 +62,7 @@ std::unique_ptr<mfem::ParMesh> ReadMesh(MPI_Comm comm, const IoData &iodata, boo
 {
   // On root, read the serial mesh (converting format if necessary), and do all necessary
-  // serial preprocessing. When finished, distribute the mesh to all processes. Count disk
+  // serial preprocessing. When eval, distribute the mesh to all processes. Count disk
```
Unsure of this change

hughcars replied:
Find and replace error, fixed in changes to appear.

sebastiangrimberg commented on palace/utils/geodata.cpp:
```diff
@@ -389,17 +390,475 @@ void AttrToMarker(int max_attr, const std::vector<int> &attrs, mfem::Array<int>
+namespace
+{
+
+// Compute a lexicographic comparison of Eigen Vector3d
```
Period to terminate all comments.

sebastiangrimberg commented on palace/utils/communication.hpp:
```diff
@@ -219,6 +245,24 @@ class Mpi
+  template <typename T>
+  static mpi::ValueAndRank<T> FindMin(const T val, MPI_Comm comm)
+  {
+    mpi::ValueAndRank<T> local{val, Mpi::Rank(comm)};
```
I'd propose the following changes to these wrappers to align better with the other ones and the MPI API:

```
diff --git a/palace/utils/communication.hpp b/palace/utils/communication.hpp
index 09febd99..a7df5855 100644
--- a/palace/utils/communication.hpp
+++ b/palace/utils/communication.hpp
@@ -5,6 +5,7 @@
 #define PALACE_UTILS_COMMUNICATION_HPP

 #include <complex>
+#include <vector>
 #include <fmt/format.h>
 #include <fmt/printf.h>
 #include <fmt/ranges.h>
@@ -127,31 +128,49 @@ inline MPI_Datatype DataType<bool>()
   return MPI_C_BOOL;
 }

-// Helper for returning the rank with a MIN or MAX reduction.
-template <typename T>
-struct ValueAndRank
+template <typename T, typename U>
+struct ValueAndLoc
 {
-  // The reduced value from the MINLOC or MAXLOC operation.
-  T value;
-  // The processor rank that has the value.
-  int rank;
-  // Convenience decay operators to allow use as the underlying value.
-  inline operator const T &() const { return value; }
-  inline operator T &() { return value; }
+  T val;
+  U loc;
 };

 template <>
-inline MPI_Datatype DataType<ValueAndRank<double>>()
+inline MPI_Datatype DataType<ValueAndLoc<float, signed int>>()
+{
+  return MPI_FLOAT_INT;
+}
+
+template <>
+inline MPI_Datatype DataType<ValueAndLoc<double, signed int>>()
 {
   return MPI_DOUBLE_INT;
 }

 template <>
-inline MPI_Datatype DataType<ValueAndRank<int>>()
+inline MPI_Datatype DataType<ValueAndLoc<long double, signed int>>()
+{
+  return MPI_LONG_DOUBLE_INT;
+}
+
+template <>
+inline MPI_Datatype DataType<ValueAndLoc<signed short, signed int>>()
+{
+  return MPI_SHORT_INT;
+}
+
+template <>
+inline MPI_Datatype DataType<ValueAndLoc<signed int, signed int>>()
 {
   return MPI_2INT;
 }

+template <>
+inline MPI_Datatype DataType<ValueAndLoc<signed long int, signed int>>()
+{
+  return MPI_LONG_INT;
+}
+
 }  // namespace mpi

 //
@@ -169,7 +188,6 @@ public:
   // Finalize MPI (if it has been initialized and not yet already finalized).
   static void Finalize()
   {
-
     if (IsInitialized() && !IsFinalized())
     {
       MPI_Finalize();
@@ -245,22 +263,40 @@ public:
     GlobalOp(len, buff, MPI_SUM, comm);
   }

-  // Return the rank that has the minimum over the communicator
-  template <typename T>
-  static mpi::ValueAndRank<T> FindMin(const T val, MPI_Comm comm)
+  // Global minimum with location (in-place, result is broadcast to all processes).
+  template <typename T, typename U>
+  static void GlobalMinLoc(int len, T *val, U *loc, MPI_Comm comm)
   {
-    mpi::ValueAndRank<T> local{val, Mpi::Rank(comm)};
-    GlobalOp(1, &local, MPI_MINLOC, comm);
-    return local;
+    std::vector<mpi::ValueAndLoc<T, U>> buff(len);
+    for (int i = 0; i < len; i++)
+    {
+      buff[i].val = val[i];
+      buff[i].loc = loc[i];
+    }
+    GlobalOp(len, buff.data(), MPI_MINLOC, comm);
+    for (int i = 0; i < len; i++)
+    {
+      val[i] = buff[i].val;
+      loc[i] = buff[i].loc;
+    }
   }

-  // Return the rank that has the maximum over the communicator
+  // Global maximum with location (in-place, result is broadcast to all processes).
   template <typename T>
-  static mpi::ValueAndRank<T> FindMax(const T &val, MPI_Comm comm)
+  static void GlobalMaxLoc(int len, T *val, U *loc, MPI_Comm comm)
   {
-    mpi::ValueAndRank<T> local{val, Mpi::Rank(comm)};
-    GlobalOp(1, &local, MPI_MAXLOC, comm);
-    return local;
+    std::vector<mpi::ValueAndLoc<T, U>> buff(len);
+    for (int i = 0; i < len; i++)
+    {
+      buff[i].val = val[i];
+      buff[i].loc = loc[i];
+    }
+    GlobalOp(len, buff.data(), MPI_MAXLOC, comm);
+    for (int i = 0; i < len; i++)
+    {
+      val[i] = buff[i].val;
+      loc[i] = buff[i].loc;
+    }
   }

   // Global broadcast from root.
```

sebastiangrimberg commented on .gitignore:
```diff
@@ -1,6 +1,7 @@
 build*/
 examples/**/postpro*/
 examples/**/log*/
+examples/**/*.log
```
As none of the automation in the repo generates `.log` files, I'd be against this. This just seems to be happening if you are piping `stdout`/`stderr` to file (if you just put it into a `log/` directory like for the tests you will be fine.

sebastiangrimberg commented on palace/utils/geodata.cpp:
```diff
@@ -389,17 +390,411 @@ void AttrToMarker(int max_attr, const std::vector<int> &attrs, mfem::Array<int>
-                    mfem::Vector &max)
+double BoundingBox::Area() const
+{
+  return 4 * normals[0].cross(normals[1]).norm();
```
Maybe just make these one-liners inline in the `BoundingBox` class declaration?

sebastiangrimberg commented on palace/utils/geodata.cpp:
```diff
@@ -389,17 +390,411 @@ void AttrToMarker(int max_attr, const std::vector<int> &attrs, mfem::Array<int>
+             comm);
+
+  std::vector<Eigen::Vector3d> collected_vertices;
+  if (is_root)
```
Just `(Mpi::Rank(comm) == dominant_rank)` to be consistent with above?

sebastiangrimberg commented on palace/utils/geodata.cpp:
```diff
@@ -389,17 +390,411 @@ void AttrToMarker(int max_attr, const std::vector<int> &attrs, mfem::Array<int>
+  mfem::Array<int> marker(bdr ? mesh.bdr_attributes.Max() : mesh.attributes.Max());
+  marker = 0;
+  marker[attr - 1] = 1;
+
```
Empty line consistency with `GetAxisAlignedBoundingBox` (prefer none).

sebastiangrimberg reviewed (commented):
Cool stuff! Thanks @hughcars.

hughcars commented on examples/cpw/cpw_lumped_uniform.json:
```diff
@@ -79,11 +79,12 @@
             "Attributes": [4],
-            "Direction": "+Y"
+            "Direction": [0,1,0],
+            "CoordinateSystem": "Cartesian"
```
I was opting to mix and match in here, to demonstrate all the different ways of handling the same. All the ports in here can be written most succinctly with `"Y"` variations, but I thought it had some value showing the equivalents. I'll make all the `"+Y"` back to keyword, and leave the others as `[0,-1,0]`

sebastiangrimberg replied:
I think within one example folder all should use the same syntax. That way we can `diff` across config files and see the real differences. Comparing across `cpw/` and `rings/`, for example, will give you the various ways to specify `"Direction"`.

hughcars commented on examples/rings/rings.json:
```diff
@@ -8,7 +8,8 @@
     "Mesh": "mesh/rings.msh",
-    "L0": 1.0e-6  // um
+    "L0": 1.0e-6,  // um
+    "Lc": 2e3
```
You're correct, I put this in to keep the normalization the same, particularly when I was repeatedly rotating the mesh to generate new data. Now that I am comfortable in the results, if we're ok with it I can just remove and use the automatically generated Lc.

sebastiangrimberg commented on examples/rings/mesh/mesh.jl:
```diff
@@ -2,203 +2,216 @@
-# Optionally launch GUI
-if "gui" in lowercase.(ARGS)
-    gmsh.fltk.run()
+    # gmsh.option.setNumber("Mesh.Algorithm3D", 10)
```
Understandable, but once your changes to the example are made and we have rebaselined we aren't regenerating meshes anymore, and certainly not to test against so identical meshing reruns don't matter. I would revert this change.

hughcars commented on palace/utils/geodata.hpp:
```diff
@@ -51,10 +53,55 @@ void AttrToMarker(int max_attr, const mfem::Array<int> &attrs, mfem::Array<int>
+
+// Computes a bounding box for a point cloud, assuming a distributed point cloud
+// of vertices.
+BoundingBox BoundingBoxFromPointCloud(MPI_Comm comm, std::vector<Eigen::Vector3d> vertices);
```
Done, but fyi it conflicts with making the Area and Volume methods inlines, unless they use home rolled calculations for the cross product etc. again.

hughcars commented on docs/src/config/boundaries.md:
```diff
@@ -258,8 +259,9 @@ accounts for nonzero metal thickness.
             {
-                "Attributes": [<int array>],
-                "Direction": <string>
+                "Attributes": <string> or <float array>,
```
done in e940516d7f2ed83d66ad3be819af81530b5d9c31

hughcars commented on palace/fem/lumpedelement.hpp:
```diff
@@ -59,39 +59,60 @@ class LumpedElementData
-  double l, w;    // Lumped element length and width
-
+  mfem::Vector direction;  // Cartesian vector specifying signed direction of incident field
+  mesh::BoundingBox bounding_box;  // Bounding box defining the rectangular lumped port
```
done in e940516d7f2ed83d66ad3be819af81530b5d9c31

hughcars commented on palace/utils/geodata.hpp:
```diff
@@ -51,10 +52,59 @@ void AttrToMarker(int max_attr, const mfem::Array<int> &attrs, mfem::Array<int>
+  std::array<double, 3> Lengths() const;
+  // Compute the deviation in degrees of a vector from each of the normal
+  // directions.
+  std::array<double, 3> Deviation(const std::array<double, 3> &direction) const;
```
Non inline methods for these calculations allows `lumpedelement.hpp` to be cleaner and avoid having `#include <Eigen/Dense>`

sebastiangrimberg commented on examples/rings/rings.json:
```diff
@@ -79,7 +80,7 @@
       "Type": "AMS",
       "KSPType": "CG",
-      "Tol": 1.0e-8,
+      "Tol": 1.0e-9,
```
This can probably be reverted now that the rings example CI is rebaselined correctly and passes.

sebastiangrimberg replied:
Done in [sjg/lumped-port-orientation-fixes](https://github.com/awslabs/palace/tree/sjg/lumped-port-orientation-fixes).

sebastiangrimberg commented on palace/utils/configfile.hpp:
```diff
@@ -60,6 +61,29 @@ struct DataMap
   [[nodiscard]] auto end() { return mapdata.end(); }
 };
 
+// A ElementData consists of a list of attributes making up a single
```
"An ElementData" or "A NodeData"

sebastiangrimberg commented:
I'm pushed a branch [sjg/lumped-port-orientation-fixes](https://github.com/awslabs/palace/tree/sjg/lumped-port-orientation-fixes) which contains some final fixes (almost all style/formatting) to this branch. If it looks good to you, feel free to merge it here before this gets merged.

sebastiangrimberg commented on palace/utils/geodata.cpp:
```diff
@@ -1193,7 +1642,7 @@ std::unique_ptr<mfem::ParMesh> DistributeMesh(MPI_Comm comm,
         MPI_Isend(so[i].c_str(), ilen, MPI_CHAR, i, i, comm, &send_requests[i - 1]);
       }
-      std::istringstream fi(so[0]);  // This is never compressed
+      std::istringstream fi(so[0]);  // This is never compressed.
```
Done in [sjg/lumped-port-orientation-fixes](https://github.com/awslabs/palace/tree/sjg/lumped-port-orientation-fixes).

sebastiangrimberg commented on test/ref/rings/terminal-M.csv:
```diff
@@ -1,3 +1,3 @@
             i,             M[i][1] (H),             M[i][2] (H)
- 1.000000e+00,        +4.258130528e-11,        +1.958445479e-12
- 2.000000e+00,        +1.958445479e-12,        +7.126915567e-10
+ 1.000000e+00,        +4.277684200e-11,        +1.959581140e-12
```
This is done in [sjg/lumped-port-orientation-fixes](https://github.com/awslabs/palace/tree/sjg/lumped-port-orientation-fixes).

sebastiangrimberg reviewed (approved):
Looks great, this is a great feature to have. Will mark as ready once https://github.com/awslabs/palace/pull/47 is merged, and will rebase and merge from there.

sebastiangrimberg commented:
@hughcars, I've rebased this on `main` and it should be ready now. Can you please check out the final commit b9c26d714d230c3800d7219d12ecf186763f1554 which I added to reduce some duplicate code between the `BoundingBoxFromPointCloud` and `BoundingBallFromPointCloud` methods? If you're OK with the changes there I will go ahead and mark this for merge.

hughcars commented on palace/utils/geodata.cpp:
```diff
@@ -560,13 +560,12 @@ bool EigenLE(const Eigen::Vector3d &x, const Eigen::Vector3d &y)
-std::vector<Eigen::Vector3d> CollectPointCloud(mfem::ParMesh &mesh,
-                                               const mfem::Array<int> &marker, bool bdr)
+// bounding balls. Returns the vector of vertices on the dominant rank, and none on all
+// ranks. Vertices are de-duplicated to a certain floating point precision.
```
```
// Helper for collecting a point cloud from a mesh, used in calculating bounding boxes and
// bounding balls. Returns the dominant rank, for which the vertices argument will be
// filled, while all other ranks will have an empty vector. Vertices are de-duplicated to a
// certain floating point precision.
```

sebastiangrimberg replied:
Fixed in [fa13b91](https://github.com/awslabs/palace/pull/75/commits/fa13b9127936f410ec4f4ddb599c29abac62e848).

hughcars commented on palace/utils/geodata.cpp:
```diff
@@ -819,85 +822,29 @@ BoundingBox BoundingBoxFromPointCloud(MPI_Comm comm, std::vector<Eigen::Vector3d
 // Calculates a bounding ball from a point cloud, result is broadcast across all processes.
 BoundingBall BoundingBallFromPointCloud(MPI_Comm comm,
-                                        std::vector<Eigen::Vector3d> vertices)
+                                        std::vector<Eigen::Vector3d> &vertices,
```
`const`

sebastiangrimberg replied:
Fixed in [fa13b91](https://github.com/awslabs/palace/pull/75/commits/fa13b9127936f410ec4f4ddb599c29abac62e848).

hughcars commented on palace/utils/geodata.cpp:
```diff
@@ -689,43 +682,52 @@ BoundingBox BoundingBoxFromPointCloud(MPI_Comm comm, std::vector<Eigen::Vector3d
+}
 
+// Calculates a bounding box from a point cloud, result is broadcast across all processes.
+BoundingBox BoundingBoxFromPointCloud(MPI_Comm comm, std::vector<Eigen::Vector3d> &vertices,
```
`const` on the vector

sebastiangrimberg replied:
Fixed in [fa13b91](https://github.com/awslabs/palace/pull/75/commits/fa13b9127936f410ec4f4ddb599c29abac62e848).

hughcars commented:
> @hughcars, I've rebased this on `main` and it should be ready now. Can you please check out the final commit [b9c26d7](https://github.com/awslabs/palace/commit/b9c26d714d230c3800d7219d12ecf186763f1554) which I added to reduce some duplicate code between the `BoundingBoxFromPointCloud` and `BoundingBallFromPointCloud` methods? If you're OK with the changes there I will go ahead and mark this for merge.

Looks good, slight change to the comment wording, and two missing `const` is all.
